pkg/proxy/proxy.go

```diff
@@ -0,0 +1,327 @@
+	"github.com/op/go-logging"
+	"github.com/spf13/viper"
+	"github.com/vulcand/oxy/forward"
+	"github.com/vulcand/route"
+)
+
+var (
+	log     = logging.MustGetLogger("cilium-proxy")
+	logFile *os.File
+	logBuf  *bufio.Writer
+)
+
+type Redirect struct {
+	id       string
+	FromPort uint16
+	ToPort   uint16
+	Rules    []policy.AuxRule
+	source   ProxySource
+	server   *manners.GracefulServer
+	router   route.Router
+}
+
+func (r *Redirect) updateRules(rules []policy.AuxRule) {
+	for _, v := range r.Rules {
+		r.router.RemoveRoute(v.Expr)
+	}
+
+	r.Rules = make([]policy.AuxRule, len(rules))
+	copy(r.Rules, rules)
+
+	for _, v := range r.Rules {
+		r.router.AddRoute(v.Expr, v)
+	}
+}
+
+type ProxySource interface {
+}
+
+type Proxy struct {
+	// rangeMin is the minimum port used for proxy port allocation
+	rangeMin uint16
+
+	// rangeMax is the maximum port used for proxy port allocation.
+	// If port is unspecified, the proxy will automatically allocate
+	// ports out of the rangeMin-rangeMax range.
+	rangeMax uint16
+
+	// nextPort is the next available proxy port to use
+	nextPort uint16
+
+	// allocatedPorts is a map of all allocated proxy ports pointing
+	// to the redirect rules attached to that port
+	allocatedPorts map[uint16]*Redirect
+
+	// redirects is a map of all redirect configurations indexed by
+	// the redirect identifier
+	redirects map[string]*Redirect
+
+	// mutex is the lock required when modifying any proxy datastructure
+	mutex sync.RWMutex
+}
+
+func NewProxy(minPort uint16, maxPort uint16) *Proxy {
+	return &Proxy{
+		rangeMin:       minPort,
+		rangeMax:       maxPort,
+		nextPort:       minPort,
+		redirects:      make(map[string]*Redirect),
+		allocatedPorts: make(map[uint16]*Redirect),
+	}
+}
+
+func (p *Proxy) allocatePort() (uint16, error) {
+	port := p.nextPort
+
+	for {
+		resPort := port
+		port++
+		if port >= p.rangeMax {
+			port = p.rangeMin
+		}
+
+		if _, ok := p.allocatedPorts[resPort]; !ok {
+			return resPort, nil
+		}
+
+		if port == p.nextPort {
+			return 0, fmt.Errorf("no available proxy ports")
+		}
+	}
+}
+
+func generateURL(w http.ResponseWriter, req *http.Request, dport uint16) (*url.URL, error) {
+	ip, port, err := net.SplitHostPort(req.RemoteAddr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid remote address: %s", err)
+	}
+
+	pIP := net.ParseIP(ip)
+	if pIP == nil {
+		return nil, fmt.Errorf("unable to parse IP %s", ip)
+	}
+
+	sport, err := strconv.ParseUint(port, 10, 16)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse port string: %s", err)
+	}
+
+	key := &Proxy4Key{
+		SPort:   uint16(sport),
+		DPort:   dport,
+		Nexthdr: 6,
+	}
+
+	copy(key.SAddr[:], pIP.To4())
+
+	log.Debugf("Looking up proxy %+v\n", key)
+
+	val, err := LookupEgress4(key)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to find proxy entry for %s: %s", key, err)
+	}
+
+	log.Debugf("Found proxy entry: %+v\n", val)
+
+	newUrl := fmt.Sprintf("http://%s%s", val.HostPort(), req.URL)
```

Comment:
The end result will be `172.0.0.1:8080cilium/clium`, isn't there a missing `/` (`"http://%s/%s"`)?